arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2017 Linaro Limited.
+ * Copyright (c) 2018 Nordic Semiconductor ASA.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* This internal function performs MPU region initialization.
+ *
+ * Note:
+ *   The caller must provide a valid region index.
+ */
+static void _region_init(u32_t index, struct arm_mpu_region *region_conf)
+{
+	/* Select the region you want to access */
+	MPU->RNR = index;
+	/* Configure the region */
+	MPU->RBAR = (region_conf->base & MPU_RBAR_ADDR_Msk)
+				| MPU_RBAR_VALID_Msk | index;
+	MPU->RASR = region_conf->attr | MPU_RASR_ENABLE_Msk;
+	SYS_LOG_DBG("[%d] 0x%08x 0x%08x",
+		index, region_conf->base, region_conf->attr);
+}
+
+#if defined(CONFIG_USERSPACE) || defined(CONFIG_MPU_STACK_GUARD) || \
+	defined(CONFIG_APPLICATION_MEMORY)
+
+/**
+ * This internal function converts the region size to
+ * the SIZE field value of MPU_RASR.
+ */
+static inline u32_t _size_to_mpu_rasr_size(u32_t size)
+{
+	/* The minimal supported region size is 32 bytes */
```

Comment:
IMHO we should assert if region is not exactly 32, 64 ... 4G. This way we will avoid risk of setting invalid attributes on memory behind region, which could be considered as security issue. Also this change could be used a verification of our build system, as it should ensure proper location/size of the regions.

Reply:
The caller of this function is `arm_core_mpu_configure_mem_domain()`, which is invoked by kernel. So, what you ask, is either already done by design, or (if not), we need to enforce the restriction in the kernel and not in the MPU driver.

Reply:
Some MPUs (like ARMv8) does not have such restrictions. I am still opting for assert here as it will check correct usage of the driver.

Reply:
Now I am a bit confused: this code is, actually, specific to ARMv7...

Reply:
OK, now I got your point. So, in principle, adding an ASSERT here is not a big deal, so I am going to add it.

Reply:
This function does not exist for ARMv8-M.

In ARMv8-M 
- such an ASSERT does not make sence, as regions can have any size.
- we, actually, do not need such an ASSERT, because if Kernel miss-configures regions, resulting in overlap, we will get an MPU fault.

Reply:
The fault on ARMv8 will be generated only if the overlapping region will be touched (read, written or executed). In some situations, the fault might be observed very raerly.

Reply:
>In some situations, the fault might be observed very raerly.

If the fault is not observed, this actually means, we never access at all the overlapping areas, so, this is an "inactive" bug, i.e. it does not get us into any trouble.

Contrary to ARMv7-M, where misconfiguration is "allowed" by the programming model, and we end up configuring regions _not_ in the intended way (so the ASSERT, there make sense).

Now, I agree that we need to make sure that we do not configure overlapping regions, but this challenge shall be addressed "by design". It is very expensive (in run-time) to iterate over all MPU regions and check that they do not overlap with the region we attempt to configure. Instead, we need to implement "adjust region" function and call it in proper places. But this should be done in a subsequent PR.

Reply:
The checking code might be activated using dedicated Kconfig option, which will be turned on during testing only.

Reply:
Discussed offline: we will need to ensure we never configure overlapping regions, but we won't do this in this PR. I will open an issue for it.

Reply:
Issue: #8907